[GSoC] Summary for Project 6: Push-based Metrics Collection

Electronic-Waste opened the pull request:
I converted the issue https://github.com/kubeflow/katib/issues/2340 into this blog.

PTAL when you have time @kubeflow/wg-automl-leads @terrytangyuan

andreyvelich reviewed (commented):
Thank you for working on this @Electronic-Waste, and sorry for the late reply!

/assign @varodrig @hbelmiro @franciscojavierarceo @kubeflow/wg-training-leads @akgraner 
Please can you help us with the review, so we can merge this great blog post ?

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+---
+
+This summer, I gained a precious opportunity to participate in the Google Summer of Code(GSoC), in which I would contribute to Katib and fulfill a project named ["Push-based Metrics Collection in Katib"](https://www.kubeflow.org/events/gsoc-2024/#project-6-push-based-metrics-collection-for-katib) within 12 weeks. 
+Firstly, I got to know about GSoC and Kubeflow with the recommendation from the former active maintainer Ce Gao(gaocegege)'s personal blog. And deeply impressed by the idea of cloud native AI toolkits, I decided to have a dip into this area and learn some skills to enhance my competence for career life in the future.
```
```suggestion
Firstly, I got to know about GSoC and Kubeflow with the recommendation from the former active maintainer Ce Gao(gaocegege)'s personal blog. And I was deeply impressed by the idea of cloud native AI toolkits, I decided to have a dip into this area and learn some skills to enhance my competence for career life in the future.
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+## What I did in GSoC Project
+
+I raised numerous PRs for the Katib and Training-Operator project. Some of them are related to my GSoC project, and others may contribute to the completeness of UTs, simplicity of dependency package, and the compatibility of UI component, etc.
```
```suggestion
I raised numerous PRs for the Katib and Training-Operator project. Some of them are related to my GSoC project, and others may contribute to the completeness of UTs (Unit Tests), simplicity of dependency management, and the compatibility of the UI component.
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+For reference, the coding period can be rougly divided into 3 stages:
+
+1. Convert the proposal to KEP and discuss with mentors about the architecture, API design, etc. (~4 weeks)
```
```suggestion
1. Convert the proposal to a KEP and discuss the architecture, API design, etc. (~4 weeks) with the mentors
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+1. Convert the proposal to KEP and discuss with mentors about the architecture, API design, etc. (~4 weeks)
+
+2. Develop push-based metrics collection interface according to the KEP. (~8 weeks)
```
```suggestion
2. Develop a push-based metrics collection interface according to the KEP. (~8 weeks)
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+2. Develop push-based metrics collection interface according to the KEP. (~8 weeks)
+
+3. Write some examples and documentations & Present my work to Kubeflow Community. 
```
```suggestion
3. Write some examples and documentation & Present my work to the Kubeflow Community. 
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+3. Write some examples and documentations & Present my work to Kubeflow Community. 
+
+Also, I raised some issues not only to describe the problems and bugs I met during the coding period, but also to suggest the future enhancement direction for Katib and Training-Operator.
```
```suggestion
Also, I raised some issues not only to describe the problems and bugs I met during the coding period, but also to suggest the future enhancement direction for Katib and the Training-Operator.
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+Also, I raised some issues not only to describe the problems and bugs I met during the coding period, but also to suggest the future enhancement direction for Katib and Training-Operator.
+
+There is an [Github Issue](https://github.com/kubeflow/katib/issues/2340) tracks the progress of developing push-based metrics collection for katib during the GSoC coding phase. If interested in my work or Katib, you can check this issue for more details.
```
```suggestion
There is a [Github Issue](https://github.com/kubeflow/katib/issues/2340) that tracks the progress of developing the push-based metrics collection for katib during the GSoC coding phase. If you interested in my work or Katib, you can check this issue for more details.
```

franciscojavierarceo commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+2. **Dive into the Source Code**: Engineering projects nowadays are extremely complex and need much effort to understand them. The best way to get familiar with the project is to dive into the source code and run several examples.
+
+3. **Communication**: Communication is the most important thing when we collaborate with others. Expressing your idea precisely and making others understand you easily are significant skills not only in open source community but also in various scenes such as company and group works.
```
```suggestion
3. **Communication**: Communication is the most important thing when collaborating with others. Expressing your idea precisely and making others understand you easily are significant skills not only in the open source community but also in various scenarios such as at a company and in group work.
```

franciscojavierarceo reviewed (commented):
This is awesome! Thank you @Electronic-Waste!! I left some small nits but otherwise lgtm.

varodrig commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+Firstly, I got to know about GSoC and Kubeflow with the recommendation from the former active maintainer Ce Gao(gaocegege)'s personal blog. And deeply impressed by the idea of cloud native AI toolkits, I decided to have a dip into this area and learn some skills to enhance my competence for career life in the future.
+In the blog, I'll provide my personal insight into Katib, for those who are interested in cloud native, AI, and hyperparameters tuning.
+
+## Goal
```
This is great, you could also separate into Problem and Solution, just to make it more clear to the audience, but the description is very good.

Electronic-Waste replied:
SGTM:)

varodrig commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+There is an [Github Issue](https://github.com/kubeflow/katib/issues/2340) tracks the progress of developing push-based metrics collection for katib during the GSoC coding phase. If interested in my work or Katib, you can check this issue for more details.
+
+## Lessons I Learned during the Project
```
suggestions, I'd keep just : Lessons learned  as subtitle.

Electronic-Waste replied:
Nice suggestion!

varodrig commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+![](../images/2024-09-28-gsoc-2024-summary-push-based-metrics-collection/push-based-metrics-collection.png)
+
+## What I did in GSoC Project
```
Suggestions:

My contributions during the GSoC

Electronic-Waste replied:
Got it!

varodrig commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,63 @@
+
+I hold a firm belief that every small step counts, and everybody in the community is unique and of great significance. There is no doubt that our joint efforts will surely contribute to the flourishing of our Kubeflow Community, make it the world-best community managing AI lifecycle on Kubernetes, and attract much more attention from the industry. Then, more and more new comers will pour in and w ...
+
+Again, I'll continue to contribute to Kubeflow.
```
is there any call to action we could include here for the readers? 
such as learn more about kubeflow or GSCOC 2025? or any other blog post you want to link here.

Electronic-Waste replied:
What about putting the kubeflow official website and the idea list of GSoC'25 here? It might be useful for readers to know more about kubeflow and GSoC.

varodrig reviewed (commented):
Great post! thank you so much for working on this. It seems you had a fantastic experience and you are sharing your thoughts with the community. Great story. 
I left a couple of suggestions.

Electronic-Waste commented:
@franciscojavierarceo @franciscojavierarceo Thanks so much for your detailed reviews! Appreciate your patience and kindness.

I've addressed all of your comments. PTAL if you have time:)

varodrig commented:
/lgtm

franciscojavierarceo reviewed (approved):
/lgtm

andreyvelich commented on _posts/2024-09-28-gsoc-2024-summary-push-basd-metrics-collection.md:
```diff
@@ -0,0 +1,72 @@
+
+The current implementation of Metrics Collector is pull-based, raising design problems such as determining the frequency at which we scrape the metrics, performance issues like the overhead caused by too many sidecar containers, and restrictions on developing environments that must support sidecar containers and admission webhooks. And also, for data scientists, they need to pay attention to the  ...
+
+## Solution
```
@Electronic-Waste It would be also nice if you could update this information with these:
1. Briefly explain how the new push-based API works
2. Link to the Kubeflow docs for the mode information
3. Add you Notebook example: https://github.com/kubeflow/katib/blob/master/examples/v1beta1/sdk/mnist-with-push-metrics-collection.ipynb

Electronic-Waste replied:
Sure. I'll add these.
